File: app/src/main/java/com/example/solar_alarm/DisplayModels/DisplayModelRepository.java
package com.example.solar_alarm.DisplayModels;

import android.app.Application;
import android.os.Build;

import androidx.annotation.NonNull;
import androidx.annotation.RequiresApi;
import androidx.lifecycle.LifecycleOwner;
import androidx.lifecycle.LiveData;
import androidx.lifecycle.Observer;

import com.example.solar_alarm.Data.Repositories.SolarAlarmRepository;
import com.example.solar_alarm.Data.Tables.SolarAlarm;

import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.locks.Lock;
import java.util.concurrent.locks.ReentrantLock;

@RequiresApi(api = Build.VERSION_CODES.O)
public class DisplayModelRepository
{
    private Application                            _Application;
    private List<SolarAlarmDisplayModel>           _SolarAlarmDisplayModels;
    private LiveData<List<SolarAlarmDisplayModel>> _LiveData;

    public DisplayModelRepository(Application application)
    {
        _Application = application;
    }

    public LiveData<List<SolarAlarmDisplayModel>> GetSolarAlarmDisplayModels()
    {
        if (_SolarAlarmDisplayModels == null)
        {
            Lock lock = new ReentrantLock();
            lock.lock();

            LiveData<List<SolarAlarm>> solarAlarms = new SolarAlarmRepository(_Application).getAll();

            _SolarAlarmDisplayModels = new ArrayList<>();

            List<SolarAlarm> values = solarAlarms.getValue();

            if (values == null)
            {
                values = new ArrayList<>();
            }

            for (SolarAlarm solarAlarm : values)
            {
                _SolarAlarmDisplayModels.add(new SolarAlarmDisplayModel(_Application, solarAlarm));
            }

            _LiveData = new LiveData<List<SolarAlarmDisplayModel>>() {
                @Override
                public void observe(@NonNull LifecycleOwner owner, @NonNull Observer<? super List<SolarAlarmDisplayModel>> _SolarAlarmDisplayModels) {
                    super.observe(owner, _SolarAlarmDisplayModels);
                }
            };

            lock.unlock();
        }

        return _LiveData;
    }

    public void Refresh()
    {
        Lock lock = new ReentrantLock();
        lock.lock();

        _SolarAlarmDisplayModels = null;

        lock.unlock();
    }
}

File: app/src/main/java/com/example/solar_alarm/DisplayModels/DisplayModelRepository.kt
package com.example.solar_alarm.DisplayModels

import androidx.annotation.RequiresApi
import android.os.Build
import android.app.Application
import com.example.solar_alarm.Data.Repositories.SolarAlarmRepository
import androidx.lifecycle.*
import java.util.ArrayList
import java.util.concurrent.locks.Lock
import java.util.concurrent.locks.ReentrantLock

@RequiresApi(api = Build.VERSION_CODES.O)
class DisplayModelRepository(private val _Application: Application) {
    private var _SolarAlarmDisplayModels: MutableList<SolarAlarmDisplayModel>? = null
    private var _LiveData: LiveData<List<SolarAlarmDisplayModel>>? = null
    fun GetSolarAlarmDisplayModels(): LiveData<List<SolarAlarmDisplayModel>>? {
        if (_SolarAlarmDisplayModels == null) {
            val lock: Lock = ReentrantLock()
            lock.lock()
            val solarAlarms = SolarAlarmRepository().all
            _SolarAlarmDisplayModels = ArrayList()
            var values = solarAlarms!!.value
            if (values == null) {
                values = ArrayList()
            }
            for (solarAlarm in values) {
                _SolarAlarmDisplayModels.add(SolarAlarmDisplayModel(_Application, solarAlarm))
            }
            _LiveData = object : LiveData<List<SolarAlarmDisplayModel?>?>() {
                override fun observe(owner: LifecycleOwner, _SolarAlarmDisplayModels: Observer<in List<SolarAlarmDisplayModel?>?>) {
                    super.observe(owner, _SolarAlarmDisplayModels)
                }
            }
            lock.unlock()
        }
        return _LiveData
    }

    fun Refresh() {
        val lock: Lock = ReentrantLock()
        lock.lock()
        _SolarAlarmDisplayModels = null
        lock.unlock()
    }
}